Add Key.donor_id property, add rep command

dannyhpy opened the pull request:
I didn't test it yet

dannyhpy committed:
Add Key.donor_id property, add rep command

dannyhpy committed:
commands.rep: Fix donor_id being null

dannyhpy commented:
**Node.js v16.3.0**

```js
src/commands/setup.ts(3,3): error TS2305: Module '"discord.js"' has no exported member 'InteractionCollector'.
src/commands/setup.ts(41,16): error TS2551: Property 'setCustomId' does not exist on type 'MessageButton'. Did you mean 'setCustomID'?
src/commands/setup.ts(45,16): error TS2551: Property 'setCustomId' does not exist on type 'MessageButton'. Did you mean 'setCustomID'?
src/commands/setup.ts(55,27): error TS2345: Argument of type 'Message | APIMessage' is not assignable to parameter of type 'MessageResolvable | "@original"'.
src/commands/setup.ts(56,34): error TS2551: Property 'customId' does not exist on type 'MessageComponentInteraction'. Did you mean 'customID'?
src/commands/setup.ts(95,16): error TS2551: Property 'setCustomId' does not exist on type 'MessageButton'. Did you mean 'setCustomID'?
src/commands/setup.ts(106,33): error TS2345: Argument of type 'Message | APIMessage' is not assignable to parameter of type 'MessageResolvable | "@original"'.
src/core/GiveawayManager.ts(1,49): error TS2305: Module '"discord.js"' has no exported member 'InteractionCollector'.
src/core/GiveawayManager.ts(107,18): error TS7006: Parameter 'button' implicitly has an 'any' type.
src/core/GiveawayManager.ts(112,33): error TS2345: Argument of type 'Message | APIMessage' is not assignable to parameter of type 'MessageResolvable | "@original"'.
  Type 'APIMessage' is not assignable to type 'MessageResolvable | "@original"'.
    Type 'APIMessage' is missing the following properties from type 'Message': patch, applicationID, channel, cleanContent, and 37 more.
src/core/GiveawayManager.ts(120,22): error TS2551: Property 'customId' does not exist on type 'ButtonInteraction'. Did you mean 'customID'?
src/core/GiveawayManager.ts(123,31): error TS2551: Property 'customId' does not exist on type 'ButtonInteraction'. Did you mean 'customID'?
src/core/GiveawayManager.ts(128,78): error TS2551: Property 'customId' does not exist on type 'ButtonInteraction'. Did you mean 'customID'?
src/core/utils.ts(3,12): error TS2305: Module '"discord.js"' has no exported member 'InteractionCollector'.
src/core/utils.ts(26,10): error TS2551: Property 'setCustomId' does not exist on type 'MessageButton'. Did you mean 'setCustomID'?
src/core/utils.ts(55,14): error TS2551: Property 'setCustomId' does not exist on type 'MessageButton'. Did you mean 'setCustomID'?
src/core/utils.ts(106,28): error TS7006: Parameter 'collected' implicitly has an 'any' type.
```

Bluenix2 commented:
Just as an update..

I am working on pin-pointing a commit in the `package.json` file so that we don't get these issues. Currently waiting on a couple of pull requests.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+  description = 'Either give a good or a bad reputation point to an user.';
+  permissions: Array<PermissionString> = ['SEND_MESSAGES'];
+  ignoreGuilds = true;
+  options = [
+    {
+      type: 'STRING' as const,
+      name: 'rep',
+      description: 'Use the words "good" or "bad"',
+      required: true
+    }
+  ];
```
I haven't looked *too* hard into it, but could we perhaps use subcommands for this?

Bluenix2 commented on src/database/GiveawayTable.ts:
```diff
@@ -68,4 +70,19 @@ export default class GiveawayTable extends Table {
   }
+
+  /**
+   * Get the latest won giveaway by a certain user.
```
```suggestion
   * Get the latest giveaway the user won.
```

Bluenix2 commented on src/database/GiveawayTable.ts:
```diff
@@ -68,4 +70,19 @@ export default class GiveawayTable extends Table {
+   * Get the latest won giveaway by a certain user.
+   * @param id The ID of the user
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The latest found giveaway, or null
```
```suggestion
   * @returns The latest giveaway found, or null
```

Bluenix2 commented on src/database/GiveawayTable.ts:
```diff
@@ -68,4 +70,19 @@ export default class GiveawayTable extends Table {
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The latest found giveaway, or null
+   */
+  async getLastWonGiveawayBy(id: Snowflake, connection?: PoolClient): Promise<Giveaway | null> {
```
```suggestion
  async getLastWon(id: Snowflake, connection?: PoolClient): Promise<Giveaway | null> {
```

Bluenix2 commented on src/database/GiveawayTable.ts:
```diff
@@ -68,4 +70,19 @@ export default class GiveawayTable extends Table {
+   */
+  async getLastWonGiveawayBy(id: Snowflake, connection?: PoolClient): Promise<Giveaway | null> {
+    const res = await (connection || this.database).query({
+      name: 'GiveawayTable_getLastWonGiveawayBy',
```
```suggestion
      name: 'GiveawayTable_getLastWon',
```

Bluenix2 commented on src/database/KeyTable.ts:
```diff
@@ -31,10 +33,10 @@ export default class KeyTable extends Table {
     const res = await (connection || this.database).query({
       name: 'KeyTable_new',
-      text:'INSERT INTO keys (value, message) VALUES ($1, $2) RETURNING *;',
+      text:'INSERT INTO keys (donor_id, value, message) VALUES ($1, $2) RETURNING *;',
       values: [value, message],
```
```suggestion
      text:'INSERT INTO keys (donor_id, value, message) VALUES ($1, $2, $3) RETURNING *;',
      values: [donor_id, value, message],
```

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+    if (lastKey == null) return interaction.reply("You're not allowed to use that command at the moment.");
+    const donor_id = lastKey.donor_id;
+    
+    const rep = interaction.options.get('rep')?.value as string;
+
+    // TODO: Check if the user already gave
+    // either a good or a bad reputation point
+
+    switch (rep) {
```
```suggestion
    // TODO: Check if the user already gave a reputation point

    switch (interaction.options.get('rep')?.value) {
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  user_id: bigint;
+  reputation: number;
+  donated_keys: number;
+  claimed_keys: number;
```
I don't yet have this in the giveaway loop, I'll add it when I do.
```suggestion
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+      CREATE TABLE IF NOT EXISTS users (
+        user_id BIGINT PRIMARY KEY,
+        reputation NOT NULL DEFAULT 0,
+        donated_keys NOT NULL DEFAULT 0,
+        claimed_keys NOT NULL DEFAULT 0
```
```suggestion
        donated_keys NOT NULL DEFAULT 0
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+   * Insert or update an user (upsert).
+   * @param id The ID of the user
+   * @param reputation The reputation of the user
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The new or updated User row
+   */
+  async updateRep(id: Snowflake, reputation: number, connection?: PoolClient): Promise<User> {
+    const query = (`
+      INSERT INTO users (user_id, reputation) VALUES ($1, $2)
+        ON CONFLICT (user_id) DO
+      UPDATE SET reputation = $2 RETURNING *;
+    `);
+    const res = await (connection || this.database).query({
+      name: 'UserTable_updateReputation',
+      text: query,
+      values: [id, reputation]
+    });
+    return res.rows[0] as User;
+  }
+
+  /**
+   * Increment reputation of an user (upsert).
+   * @param id The ID of the user
+   * @param reputationInc Reputation increment
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The new or updated User row
+   */
+  async incrementRep(id: Snowflake, reputationInc: number, connection?: PoolClient): Promise<User> {
+    const query = (`
+      INSERT INTO users (user_id, reputation) VALUES ($1, $2)
+        ON CONFLICT (user_id) DO
+      UPDATE SET reputation = reputation + $2 RETURNING *;
+    `);
+    const res = await (connection || this.database).query({
+      name: 'UserTable_updateReputation',
+      text: query,
+      values: [id, reputationInc]
+    });
+    return res.rows[0] as User;
+  }
```
Do we need both?

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+    return res.rows[0] as User;
+  }
+
+  /**
+   * Increment donated keys of an user (upsert).
+   * @param id The ID of the user
+   * @param donatedKeysInc Donated keys increment
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The new or updated User row
+   */
+  async incrementDonatedKeys(id: Snowflake, donatedKeysInc: number, connection?: PoolClient): Promise<User> {
+    const query = (`
+      INSERT INTO users (user_id, donated_keys) VALUES ($1, $2)
+        ON CONFLICT (user_id) DO
+      UPDATE SET donated_keys = donated_keys + $2 RETURNING *;
+    `);
+    const res = await (connection || this.database).query({
+      name: 'UserTable_incrementDonatedKeys',
+      text: query,
+      values: [id, donatedKeysInc]
+    });
+    return res.rows[0] as User;
+  }
```
Why have an increment number? Surely we will only ever increment it by 1.

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * Get an user.
```
```suggestion
   * Get a user by its ID.
```

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+  name = 'rep';
+  description = 'Either give a good or a bad reputation point to an user.';
+  permissions: Array<PermissionString> = ['SEND_MESSAGES'];
+  ignoreGuilds = true;
```
I changed this because it's kind of backwards. It should be `DMOnly`.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+  async run(interaction: CommandInteraction): Promise<void> {
+    const conn = await interaction.client.db.connect();
+    
+    const lastGiveaway = await interaction.client.db.giveaways.getLastWonGiveawayBy(interaction.user.id);
+    if (lastGiveaway == null) return interaction.reply("You're not allowed to use that command at the moment.");
+    const lastKey = await interaction.client.db.keys.get(lastGiveaway.key)
+    if (lastKey == null) return interaction.reply("You're not allowed to use that command at the moment.");
```
Can we add some spaces here? It's a bit unreadable right now

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+    if (lastGiveaway == null) return interaction.reply("You're not allowed to use that command at the moment.");
+    const lastKey = await interaction.client.db.keys.get(lastGiveaway.key)
+    if (lastKey == null) return interaction.reply("You're not allowed to use that command at the moment.");
+    const donor_id = lastKey.donor_id;
```
Kind of uneccessary shortcut.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,44 @@
+  ];
+
+  async run(interaction: CommandInteraction): Promise<void> {
+    const conn = await interaction.client.db.connect();
```
You never release this connection

Bluenix2 reviewed (commented):
Great work so far Danny!

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+ * An inserted User row
+ */
+export interface User {
+  user_id: bigint;
```
As we agreed this should be a discord.js Snowflake

Bluenix2 commented on src/database/KeyTable.ts:
```diff
@@ -1,9 +1,11 @@
 
 export interface Key {
   id: number;
+  donor_id: bigint;
```
This should be typed as a discord.js Snowflake as well

dannyhpy committed:
More clarity based on the review

dannyhpy committed:
rep: Use SUB_COMMAND options + one use allowed

- Use SUB_COMMAND options implemented in 32bf698418e353a0ca0aa3b5c6dd2d636d817c0f

- Prevent the user from giving a reputation point if he has already given one previously

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,62 @@
+
+export default class RepCommand extends Command {
+  name = 'rep';
+  description = 'Either give a good or a bad reputation point to an user.';
```
I think there's improvements to be made on this, perhaps just simplifying it to:
```suggestion
  description = 'Give the donator of your last won giveaway a reputation point.';
```

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,62 @@
+      type: 'SUB_COMMAND' as const,
+      name: 'good',
+      description: 'Give a good reputation point',
+      required: true
```
If I am not mistaken subcommands cannot have `required`.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    const lastKey = await interaction.client.db.keys.get(lastGiveaway.key)
+    if (lastKey == null) return interaction.reply("No key was found on your latest giveaway.");
```
Could you add a comment here that this will never happen, but we do it for the typing.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,62 @@
+    return interaction.reply('A good reputation point has been given.');
+  }
+
+  async bad(interaction: CommandInteraction): Promise<void> {
```
These two methods do basically the same thing, can you create a "middle-man" method (perhaps called `give` or similar). It takes the connection, interaction, and what increment for the rep - and will grab the last key and giveaway like this code does.

Bluenix2 commented on src/database/GiveawayTable.ts:
```diff
@@ -69,4 +72,27 @@ export default class GiveawayTable extends Table {
+    return res ? res.rows[0] as Giveaway : null;
+  }
+
+  async setRepGiven(id: number, given: boolean, connection?: PoolClient): Promise<void> {
```
Can you add a JSDoc here?

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+  
+  /**
+   * Insert or update an user (upsert).
```
```suggestion
   * Insert or update a user (upsert).
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * Increment reputation of an user (upsert).
```
```suggestion
   * Increment reputation of a user (upsert).
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+   * @param connection The connection to use, defaults to a new connection from the pool
+   * @returns The new or updated User row
+   */
+  async incrementRep(id: Snowflake, reputationInc: number, connection?: PoolClient): Promise<User> {
```
It should be safe to assume that the user is already in the database (they have claimed enough keys to be able to donate) so we may not need to upsert. Besides, I am not sure if the update will work? Hopefully `reputation + $2` takes the form of `0 + $2` because of our default.

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * Increment donated keys of an user (upsert).
```
```suggestion
   * Increment donated keys of a user (upsert).
```

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,107 @@
+  /**
+   * Increment donated keys of an user (upsert).
+   * @param id The ID of the user
+   * @param donatedKeysInc Donated keys increment
```
```suggestion
   * @param donatedKeysInc How much the increment the donated keys value
```

dannyhpy committed:
rep: use a middleware func to avoid duplicate code

- change <RepCommand>.description

- remove <RepCommand>[option].required properties

- add method <RepCommand>.give() to remove duplicated code

- <UserTable>.incrementRep() is no longer an upsert

- add JSDoc to <GiveawayTable>.setRepGiven

dannyhpy committed:
rep: Wrap give() method into a transaction

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -22,20 +22,27 @@ export default class RepCommand extends Command {
+    try {
+      const lastGiveaway = await interaction.client.db.giveaways.getLastWon(interaction.user.id);
+      if (lastGiveaway == null) return interaction.reply("You cannot use this command at the moment.");
+      if (lastGiveaway.rep_given === true) {
```
I am not completely convinced, but maybe make this a `else if`?

dannyhpy replied:
I can but does it really change anything?

Bluenix2 replied:
Not really no, but I guess it's more about readability. Since we know that the logic should go `if`, and if that failed then check that condition.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -22,20 +22,27 @@ export default class RepCommand extends Command {
+      if (lastKey == null) return interaction.reply("No key was found on your latest giveaway.");
+
+      await interaction.client.db.users.incrementRep(lastKey.donor_id, increment, conn);
+      await interaction.client.db.giveaways.setRepGiven(lastGiveaway.id, true, conn);
```
The more I think about - why not combine these functions into one. Have a method on the user table that takes an id and increments rep as well as sets that check. I think something like the following query should work:
```sql
WITH (
    SELECT id, key FROM giveaways WHERE winner = $1 ORDER BY timestamp DESC;
) AS last_giveaway;
UPDATE giveaways SET rep_given = TRUE WHERE id = last_giveaway.id;
UPDATE users SET reputation = reputation + $2 WHERE user_id = (
    SELECT donor_id FROM keys WHERE id = last_giveaway.key;
) RETURNING *;
```
You may skip that `with` statement and just pass the lastGiveaway variable.

dannyhpy committed:
Add a specific query to RepCommand

- Delete <UserTable>.incrementRep() since it was no longer used
- Delete <GiveawayTable>.setRepGiven() since it was no longer used

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -27,16 +27,28 @@ export default class RepCommand extends Command {
+      await conn.query({
+        name: 'RepCommand_giveReputationPoint',
+        text: `
+          WITH (
+            SELECT id, key FROM giveaways WHERE winner = $1 ORDER BY timestamp DESC;
+          ) AS last_giveaway;
```
You already have the `lastGiveaway` const. You can skip this and pass that (you will need to change the parameters.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,71 @@
+
+    try {
+      const lastGiveaway = await interaction.client.db.giveaways.getLastWon(interaction.user.id);
+      if (lastGiveaway == null) return interaction.reply("You cannot use this command at the moment.");
```
Something I've noticed is that you're a bit inconsistent with which quotes to use. I think I've so far tried to use single-quotes (mostly stemming from my Python background probably). Do you prefer double-quotes/see it more often?

maxerbox replied:
I think @dannyhpy should use your eslint rc file

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,71 @@
+    }
+  ];
+
+  async give(interaction: CommandInteraction, increment: number): Promise<void> {
```
Might be worth it to add a small JSDoc explaining why we made this?

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  async good(interaction: CommandInteraction): Promise<void> {
+    await this.give(interaction, 1)
+    
+    return interaction.reply('A good reputation point has been given.');
```
Currently this will both send a failure message and a `'A good reputation point has been given.'` even when it fails.

You should make `give()` return either undefined (so that we can keep the one-liners) or true (if it was successful). Then you can use this return-value to see if you should send a confirmation message.

Bluenix2 commented on src/commands/rep.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  async bad(interaction: CommandInteraction): Promise<void> {
+    await this.give(interaction, -1)
+    
+    return interaction.reply('A bad reputation point has been given.');
```
See my comment on `good()`.

Bluenix2 commented on src/database/KeyTable.ts:
```diff
@@ -31,11 +33,11 @@ export default class KeyTable extends Table {
       name: 'KeyTable_new',
-      text:'INSERT INTO keys (value, message) VALUES ($1, $2) RETURNING *;',
-      values: [value, message],
+      text:'INSERT INTO keys (donor_id, value, message) VALUES ($1, $2, $3) RETURNING *;',
```
Oops, there should be a space before the `'`.

Bluenix2 commented on src/database/UserTable.ts:
```diff
@@ -0,0 +1,86 @@
+  user_id: Snowflake;
+  reputation: number;
+  donated_keys: number;
+  claimed_keys: number;
```
Do you want to make the necessary change to have this work?

You will have to change some things inside `core/GiveawayManager.ts`.

Bluenix2 commented:
@dannyhpy can you finish this or do you perhaps want me to take over?

dannyhpy commented:
> @dannyhpy can you finish this or do you perhaps want me to take over?

@Bluenix2 Sure, feel free to take over

Bluenix2 commented:
@dannyhpy is there anything else you rather work on, or do you not really have time?
